graylog2-server/src/main/java/org/graylog2/shared/plugins/PluginLoader.java

```diff
@@ -99,9 +100,24 @@ public PluginLoader(File pluginDir, ChainingClassLoader classLoader) {
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
-        // Create one URL class loader for all plugins so they can see each other.
-        // This makes plugin inter-dependencies work.
-        classLoader.addClassLoader(URLClassLoader.newInstance(urls.toArray(new URL[urls.size()])));
+        final List<URL> sharedClassLoaderUrls = new ArrayList<>();
+        urls.forEach(url -> {
+            final PluginProperties properties = PluginProperties.fromJarFile(url.getFile());
+
+            // Decide whether to create a separate, isolated class loader for the plugin. When the plugin is isolated
+            // (the default), it gets its own class loader and cannot see other plugins. Plugins which are not
+            // isolated share one class loader so they can see each other. (makes plugin inter-dependencies work)
+            if (properties.isIsolated()) {
+                LOG.debug("Create isolated class loader for {}", url);
+                classLoader.addClassLoader(URLClassLoader.newInstance(new URL[]{url}));
+            } else {
+                LOG.debug("Use shared class loader for {}", url);
+                sharedClassLoaderUrls.add(url);
+            }
+        });
+
+        LOG.debug("Create shared class loader for {} plugins: {}", sharedClassLoaderUrls.size(), sharedClassLoaderUrls);
+        classLoader.addClassLoader(URLClassLoader.newInstance(sharedClassLoaderUrls.toArray(new URL[sharedClassLoaderUrls.size()])));
```

Comment:
Good point. If there are no plugins which request a shared class loader, we should not add an empty one.